### ISSUE TITLE
Port redirection issues

### ISSUE THREAD

**Issue by licaon-kter** [M66B/NetGuard#345]
Set a custom DNS and then change connection status (disconnect/Wi-Fi/mobile).

**Comment by M66B** [comment]
Before switch to mobile:

```
02-25 09:29:30.271  4532  4622 I NetGuard.Service: DNS system=fd00::be05:43ff:fe8c:aaff,192.168.178.1 VPN=8.8.8.8
02-25 09:29:30.271  4532  4622 I NetGuard.Service: dns=/8.8.8.8
02-25 09:29:30.272  4532  4622 I NetGuard.Service: dns=/fd00::be05:43ff:fe8c:aaff
02-25 09:29:30.272  4532  4622 I NetGuard.Service: dns=/192.168.178.1
```
After switching to mobile:
```
02-25 09:31:06.413  4532  4622 I NetGuard.Service: DNS system=194.151.228.2,194.151.228.50 VPN=8.8.8.8
02-25 09:31:06.413  4532  4622 I NetGuard.Service: dns=/8.8.8.8
02-25 09:31:06.414  4532  4622 I NetGuard.Service: dns=/194.151.228.2
02-25 09:31:06.414  4532  4622 I NetGuard.Service: dns=/194.151.228.50
```
So, it works as expected.

Note that the behavior was changes in one of the later beta's, which I described in the release announcement. If you want to use a single DNS address, you need to forward port 53 to the DNS server.

I have changed the behavior, because I had trouble myself with only one DNS address, because my first DNS address at home is over IPv6, while the IPv6 is not 100% reliable.

**Comment by licaon-kter** [comment]
In your snippets you don't have any custom servers or?

I'm not seeing this though.

Steps:
* custom DNS set up
* port redirection 53
* https://dnsleaktest.com -> Extended test -> 1 correct server
* switch connection
* custom DNS got erased
* port redirection 53 still in place
* https://dnsleaktest.com -> Extended test -> 3 wrong servers from ISP
* change custom DNS again
* https://dnsleaktest.com -> 1 correct server

**Comment by M66B** [comment]
I had set 8.8.8.8 as custom server, which is properly at the top of the list each time.

What do you mean by '*custom DNS got erased*'? 
Where is it erased? From the settings?

**Comment by licaon-kter** [comment]
Ok, so your 8.8.8.8 stays there, mine does not.
Yes, the custom DNS entry is empty.

**Comment by M66B** [comment]
Can't be, since there is just one place where the settings is written and that is in the settings window.
```
marcel@C720:~/Documents/android/projects/NetGuard$ grep -R \"dns\".*apply app/src/main
app/src/main/java/eu/faircode/netguard/ActivitySettings.java:                prefs.edit().remove("dns").apply();
```

**Comment by licaon-kter** [comment]
Umm ok, retrying this yields something else than I though....

Steps:
* open Settings
* custom DNS is current DNS from Wi-Fi
* open custom DNS
* see gray shadow of the current DNS from Wi-Fi
* write a DNS of choice
* press ok
* custom DNS is STILL shown as current DNS from Wi-Fi
* open custom DNS again right away
* STILL see gray shadow of the current DNS from Wi-Fi

What happens to the custom one?
Shouldn't it be shown in Settings after I change it? 
Shouldn't it be shown upon entering the custom DNS dialog again? 

Now, turn of Wi-Fi, it switches to mobile, redo steps... reopen dialog right away to see my custom DNS.
Exit Settings, pull to refresh/wait for hourglass to disappear, re enter Settings, reopen dialog to see my custom DNS has gone now.

**Comment by M66B** [comment]
"*custom DNS is STILL shown as current DNS from Wi-Fi*" then the address you entered was invalid.
Invalid address will be rejected to prevent problems later.

**Comment by licaon-kter** [comment]
127.0.0.1 could be invalid in what way?

**Comment by M66B** [comment]
Local addresses cannot be used for DNS servers, since the Android VPN service will not accept it.
You can use the local address by settings a port forward.

**Comment by licaon-kter** [comment]
So these logcat https://gist.github.com/licaon-kter/ab56c9cd44e0f54dd97b lines are ok then?
```
NetGuard.Log: Preference dns=null
NetGuard.Main: Preference dns=null
```

Not to mention that DNS is run as `root` and port forwarding is done to `nobody`, although it works, when it works.

**Comment by M66B** [comment]
This is okay, since NetGuard erases the DNS entry, because it was invalid.

**Comment by licaon-kter** [comment]
Port redirection is not really working all the time here.

Wi-Fi DNS is OpenDNS. Port redirection 53 to 127.0.0.1:53 user `root` (it's started by user root, does it matter?)
Via Termux as user:

```
$ dig somedomain.com
snip response
;; SERVER: 8.8.8.8#53(8.8.8.8)
```

Where does `8.8.8.8` come from?

Anyway, switching connections, DNS servers popup from either Wi-Fi or mobile.

I can check with:
https://ipleak.net or https://dnsleaktest.com (multiple DNS servers)
http://dnssec.vs.uni-due.de (not validating DNSSEC signature while my local one does)
https://dnssec-tools.org (DNSSEC not secured)
https://www.grc.com/dns/dns.htm (multiple DNS servers; not validating DNSSEC signature while my local one does)

I need to test, but it kinda works as expected when rebooting, but once I switch (off) connections it's not working anymore.
Multiple IPs are not an issue, if the connections are really redirected, but multiple DNS IPs come and go (Google/OpenDNS have many) and I'm not sure they're cached, so even if they are, since they're not validating DNSSEC means the device it's actually talking with *those* servers and not my own.

**Comment by M66B** [comment]
On opening a socket, which will be used to forward traffic to the internet, there will be a warning log line in the logcat when the port is forwarded.

https://github.com/M66B/NetGuard/blob/master/app/src/main/jni/netguard/udp.c#L405
https://github.com/M66B/NetGuard/blob/master/app/src/main/jni/netguard/tcp.c#L827

Maybe it can help to figure out what is happing.

**Comment by licaon-kter** [comment]
I do get `UDP4 redirect to...` on Wi-Fi, so turn Wi-Fi off and then on, https://dnsleaktest.com/ caught 3 servers, 2 from OpenDNS from Wi-Fi config and my custom one.
Retrying later only yielded my server, so this is working as intended.

Snippet of log:
```
NetGuard.JNI: UDP close from 10.1.10.1/57222 to wifidns/53 socket 354
NetGuard.JNI: sessions ICMP 0 UDP 3 TCP 6 max 9/3276 idle 0 sdk 22
NetGuard.JNI: pselect ready 1
NetGuard.JNI: get uid v4 p17 10.1.10.1/33050
NetGuard.JNI: Packet v4 10.1.10.1/33050 > wifidns/53 proto 17 flags  uid 0
NetGuard.JNI: UDP new session from 10.1.10.1/33050 to wifidns/53
NetGuard.JNI: UDP socket 354
NetGuard.JNI: qname www.example.com
NetGuard.JNI: DNS query qtype 1 qclass 1 name www.example.com
NetGuard.JNI: UDP forward from tun 10.1.10.1/33050 to wifidns/53 data 33
NetGuard.JNI: UDP4 redirect to 127.0.0.1/53

```

**Comment by licaon-kter** [comment]
0.93-beta

Now trying the same on mobile connection falls short, those DNS addresses never change.
Sometimes I get a "UDP4 redirect to..." but mostly:

```
NetGuard.JNI: pselect ready 1
NetGuard.JNI: UDP recv bytes 191 from telcodns/53 for tun
NetGuard.JNI: DNS response qcount 1 acount 1
NetGuard.JNI: qname www.example.com
NetGuard.JNI: DNS question 0 qtype 1 qclass 1 qname www.example.com
NetGuard.JNI: DNS qname compression ptr 12 len 6
NetGuard.JNI: qname www.example.com
NetGuard.JNI: qname www.example.com DNS answer 0 qname  1 ttl 3600 data 134.91.78.141
NetGuard.JNI: UDP sending to tun 357 from 81.12.128.206/53 to 10.1.10.1/37209 data 219
NetGuard.JNI: Loop thread b502b000
NetGuard.JNI: UDP close from 10.1.10.1/37209 to telcodns/53 socket 365
NetGuard.JNI: sessions ICMP 0 UDP 1 TCP 23 max 24/3276 idle 0 sdk 22
```

**Comment by licaon-kter** [comment]
More findings...

(I have `root` allowed on all, `nobody` blocked on all. Does it matter? I never got a notification for `nobody` blocked on accessing the network though)

Settings port redirection 53->127.0.0.1:53 as application `root` will *never* do `UDP4 redirect to 127.0.0.1/53`no matter what connection type:
```
UDP new session from 10.1.10.1/43577 to telcoISP/53
UDP socket 263
UDP forward from tun 10.1.10.1/43577 to telcoISP/53 data 32
UDP recv bytes 128 from telcoISP/53 for tun
UDP sending to tun 874 from telcoISP/53 to 10.1.10.1/43577 data 156
UDP close from 10.1.10.1/43577 to telcoISP/53 socket 263
```

Setting it as `nobody` will do it (according to the logcat at least; see lines 4-7 below), *but* NetGuard will show a notification for application `root`(?!) as accessing the DNS:
```
UDP new session from 10.1.10.1/43577 to telcoISP/53
UDP socket 263
UDP forward from tun 10.1.10.1/43577 to telcoISP/53 data 32
UDP4 redirect to 127.0.0.1/53
UDP forward from tun 10.1.10.1/59314 to externaldns/port data 512
UDP recv bytes 432 from externaldns/port for tun
UDP sending to tun 874 from externaldns/port to 10.1.10.1/59314 data 460
UDP recv bytes 128 from telcoISP/53 for tun
UDP sending to tun 874 from telcoISP/53 to 10.1.10.1/43577 data 156
UDP close from 10.1.10.1/43577 to telcoISP/53 socket 263
```

Now, I see the snippet above on both Wi-Fi and mobile, but on mobile it does not actually work, I still get responses from actual DNS servers instead of my own.

Also these popup in the logcat when forward is set up as user `nobody`, not sure if a problem:
```
type=1400 audit(0.0:1747): avc: denied { read } for lport=59314 scontext=u:r:init:s0 tcontext=u:r:init:s0 tclass=udp_socket op_res=0 ppid=1 pcomm="init" tgid=1 tgcomm="init"
type=1400 audit(0.0:1748): avc: denied { write } for laddr=127.0.0.1 lport=53 scontext=u:r:init:s0 tcontext=u:r:init:s0 tclass=udp_socket op_res=0 ppid=1 pcomm="init" tgid=1 tgcomm="init"

```

**Comment by M66B** [comment]
Root does the DNS requests and selecting root as port forwarding application will exempt root from the port forwarding.

The access notifications are independent from port forwarding.

If you see in the logging the port is forwarded (the traffic is redirected) it is really redirected.
Forwarding is independent of the connection type.

The SELinux logging explains what is going wrong: normal applications like NetGuard may not used privileged ports (port numbers below 1024). I will change the port forwarding dialog to no accept ports below 1024.

**Comment by M66B** [comment]
I have reverted this change, since port forwarding to privileged ports of external addresses is allowed.

**Comment by licaon-kter** [comment]
> Root does the DNS requests and selecting root as port forwarding application will exempt root from the port forwarding.

Oh, forgot that, so it's as expected.

> Forwarding is independent of the connection type.

Yet not here, not working on mobile but working on Wi-Fi.

> If you see in the logging the port is forwarded (the traffic is redirected) it is really redirected (at least attempted, but SELinux, see below, may in fact prevent the traffic).

I do see the SELinux warning/error, but it's working on Wi-Fi (only).

> The SELinux logging explains what is going wrong: normal applications like NetGuard may not use privileged ports (port numbers below 1024)

NetGuard is not actually using/listening on that port, so I don't see why this should be an issue though.

**Comment by M66B** [comment]
NetGuard is actually using 127.0.0.1:53, since it opens a socket to forward traffic to this address, which is not allowed by SELinux. I have added a check for this now.

**Comment by licaon-kter** [comment]
Ok, it's ok, having this/anything on <1024 is not necessary, actually this setup will transform to a no-root one, once I get this forwarding issue solved that is.

**Comment by licaon-kter** [comment]
Any ideas on how to proceed?

**Comment by M66B** [comment]
The crash is caused by an Android problem, which I cannot solve. The Linux binder is out of memory for some reason. You could try another kernel if you have root.

licaon-kter <notifications@github.com> schreef op 26 februari 2016 16:33:17 CET:

**Comment by licaon-kter** [comment]
No, I can't change kernels since I only have root, not an unlocked bootloader. ;)

I was asking about this issue: *Port forwarding not working on mobile connection*

**Comment by M66B** [comment]
There is really no difference for port forwarding on a WiFi or mobile connection.


licaon-kter <notifications@github.com> schreef op 26 februari 2016 16:50:27 CET:

**Comment by licaon-kter** [comment]
So then, there is a problem on how those sites report?

What am I doing wrong then?

**Comment by M66B** [comment]
I don't know if or what is going wrong.

licaon-kter <notifications@github.com> schreef op 26 februari 2016 17:06:09 CET:

**Comment by licaon-kter** [comment]
BTW, setting VPN DNS does not work on mobile either. Stil using telcos DNS. :(

Do all these work for you on mobile?

**Comment by M66B** [comment]
Also here there is no difference for different connection types and yes it works for me.

licaon-kter <notifications@github.com> schreef op 26 februari 2016 20:23:28 CET:

**Comment by licaon-kter** [comment]
Say set it up as OpenDNS ( 208.67.222.222 presuming you don't use them already ) and then go to https://dnsleaktest.com/ -> Extended test -> ?

**Comment by M66B** [comment]
I cannot test now, because I am on a bad mobile connection.

Did you take the DNS cache timeout into account?

licaon-kter <notifications@github.com> schreef op 26 februari 2016 20:25:36 CET:

**Comment by licaon-kter** [comment]
I could actually be fooled that it works, but no, I guess it gets flushed maybe.
* Wi-Fi, VPN DNS: OpenDNS 
* extended test -> 3 or 4 from OpenDNS
* switch Wi-Fi off so mobile kicks in
* VPN DNS is stil OpenDNS in Settings
* extended test -> 3 from TelCo :(

Should I file a separate issue for this? `VPN DNS not applying on mobile`

**Comment by M66B** [comment]
Even if you would file a separate issue, I could do nothing else than close it with 'cantfix', since there is just one API to set the DNS address of the VPN.


licaon-kter <notifications@github.com> schreef op 27 februari 2016 07:21:30 CET:

**Comment by licaon-kter** [comment]
And it works on your Nexus so this is a ROM issue then?

**Comment by M66B** [comment]
The DNS leak test doesn't complete for me (I am on a bad edge connections, because I am in a border area), but if I set another DNS address, I will see it in the traffic log, so it is being used by the DNS.

This could very well be a (deliberate?) bug in your Android version. Which device/Android version is this? Maybe we should start to write down these incompatibilities on GitHub (applies to your tethering issue as well, which was reported by somebody else too).


licaon-kter <notifications@github.com> schreef op 27 februari 2016 08:10:55 CET:

**Comment by licaon-kter** [comment]
Sony Xperia 5.1.1 (firmware 23.4.A.1.264 and 23.4.A.1.200)

**Comment by licaon-kter** [comment]
Samsung 5.0.2 initial tests, of course here it works, just so things are always upside down, here VPN (no filtering) fails, routing tether fails, yet mobile connection has correct (read my) DNS. :(

**Comment by M66B** [comment]
Samsung is not really know for quality Android versions

licaon-kter <notifications@github.com> schreef op 28 februari 2016 10:40:44 CET:

**Comment by licaon-kter** [comment]
Some progress, both [OpenVPN for Android](https://github.com/schwabe/ics-openvpn) (with external free VPN) and [Orbot](https://gitweb.torproject.org/orbot.git) (local VPN like NetGuard; albeit redirected to Google's 8.8.8.8) work ok, as in not using TelCo's DNS.
I'll try later to rummage through their code.

**Comment by licaon-kter** [comment]
Figured it out, so (at least) [here](https://github.com/M66B/NetGuard/blob/7ee7e81d406b66adbfbbb006ec6350a5ade7cff7/app/src/main/java/eu/faircode/netguard/ActivitySettings.java#L611) (probably above for IPs) and [here](https://github.com/M66B/NetGuard/blob/4a7d8c9e29e1a5c9f1f4ace45ef5835f526ec14c/app/src/main/java/eu/faircode/netguard/ActivityForwarding.java#L101) or maybe [here](https://github.com/M66B/NetGuard/blob/b4755e1e50241bbdc746693f8b94bf9ef3a8a324/app/src/main/java/eu/faircode/netguard/ActivityForwardApproval.java#L113) the reload process is not enough.

I ended up adding an `ignore system DNS` feature (you want it?) just so I can see that: **after changing VPN DNS or adding a forwarding rule you NEED to stop and start NetGuard for the rules to apply** (no amount of hourglass starring and pull to refresh helps)

**Comment by licaon-kter** [comment]
And this more important applies to connection change (Wi-Fi to mobile and reverse).

**Comment by M66B** [comment]
This check for vital parameters of the VPN being changed and restarting it when that is the case (with a seamless handover):

https://github.com/M66B/NetGuard/blob/master/app/src/main/java/eu/faircode/netguard/SinkholeService.java#L414

'''builder.equals(last_builder)''

https://github.com/M66B/NetGuard/blob/master/app/src/main/java/eu/faircode/netguard/SinkholeService.java#L2050

Forwarding is not checked and there should be no need for this.

What exactly did you change to make it work for you?

**Comment by M66B** [comment]
Can you try if the above commit fixes your problem?

**Comment by licaon-kter** [comment]
Nothing really, my change (toggle to not even add sysdns to the VPN creation) was useless since (at least on these devices) it did not fix the need for a restart (not reload).

Basically, start on Wi-Fi, create VPN only with 8.8.8.8 as DNS, it's ok, switch to mobile (without restart) and you see TelCo DNS, restart NetGuard (not reload) and only now 8.8.8.8 is used.

**Comment by M66B** [comment]
Then the commit will likely fix this problem.

**Comment by licaon-kter** [comment]
I'll be able to compile this only later today, but guessing it will fix this case, it might need to be applied at least to VPN DNS change (IP DNS too) and port forwarding, as far as my issues here show me.

The VPN DNS and port forwarding bypass makes me think it's an Android bug, on connection change, as it won't be the first in this area.

**Comment by licaon-kter** [comment]
*NetGuard-v2.0-2016022901-all-release* not helping :(

What does `UDP idle 16/15 sec state 0 from 10.1.10.1/14533 to 208.67.222.222/53` mean? Had some of these on mobile connection after switching from Wi-Fi, where 208.67.222.222 is VPN DNS but not actually used since still TelCo DNSes would respond.

Comparing a `restart on connection change` as the app does it (and custom DNS is not working and neither does port forwarding for port 53):
```
NetGuard.Service: Native restart
NetGuard.Service: Stop native clear=false
NetGuard.JNI: Stop tun 192 clear 0 thread b502b000
NetGuard.JNI: Kill thread b502b000
NetGuard.JNI: pselect signaled tun 192 thread b502b000
NetGuard.Service: Native exit reason=null
NetGuard.JNI: Stopped events tun=192 thread b502b000
NetGuard.JNI: Join thread b502b000
NetGuard.JNI: Stopped thread b502b000
NetGuard.Service: Start native log=false filter=true
NetGuard.Service: Hosts file use=true exists=false
NetGuard.Service: Forward protocol=17 port 53 to 127.0.0.1/5353 uid 9999
NetGuard.JNI: Starting tun 192 fwd53 1 level 5 thread 0
NetGuard.JNI: Started thread b502b000
NetGuard.JNI: Start events tun=192 thread b502b000
```

vs `actually stopping and restarting` the app (and everything is working as expected):
```
NetGuard.Main: Switch=false
NetGuard.Main: Preference enabled=false
NetGuard.Service: Start intent=Intent { cmp=eu.faircode.netguard/.SinkholeService (has extras) } command=stop reason=switch off vpn=true user=0
NetGuard.Service: Executing intent=Intent { cmp=eu.faircode.netguard/.SinkholeService (has extras) } command=stop reason=switch off vpn=true user=0
NetGuard.Service: Stop native clear=true
NetGuard.JNI: Stop tun 223 clear 1 thread b502b000
NetGuard.JNI: Kill thread b502b000
NetGuard.JNI: Join thread b502b000
NetGuard.JNI: pselect signaled tun 223 thread b502b000
NetGuard.Service: Native exit reason=null
NetGuard.JNI: Stopped events tun=223 thread b502b000
NetGuard.JNI: Stopped thread b502b000
NetGuard.Service: Stopping
NetGuard.Util: Size=1 (Integer)
NetGuard.Service: Stop foreground state=enforcing
NetGuard.Util: Metered=false (Boolean)
NetGuard.Util: Connected=false (Boolean)
NetGuard.Service: Messages=false wakelock=false
NetGuard.Util: Size=0 (Integer)
NetGuard.Util: SIM=ro network=ro
NetGuard.Service: In service international=0
NetGuard.Service: Data connected generation=3G
NetGuard.Main: Switch=true
NetGuard.Main: Preference enabled=true
NetGuard.Main: Prepare done
NetGuard.Main: onActivityResult request=1 result=1 ok=true
NetGuard.Service: Start intent=Intent { cmp=eu.faircode.netguard/.SinkholeService (has extras) } command=start reason=prepared vpn=false user=0
NetGuard.Service: Executing intent=Intent { cmp=eu.faircode.netguard/.SinkholeService (has extras) } command=start reason=prepared vpn=false user=0
NetGuard.Service: Start foreground state=enforcing
NetGuard.Service: Get allowed connected=true wifi=false home="WiFi1","WiFi2" network=<unknown ssid> metered=true/true generation=3G roaming=false/false interactive=true tethering=true filter=true
NetGuard.Service: Allowed 17 of 245
NetGuard.Service: vpn4=10.1.10.1 vpn6=fd00:1:fd00:1:fd00:1:fd00:1
NetGuard.Service: DNS system=telcoDNS2,telcoDNS1 VPN=208.67.222.222
NetGuard.Service: dns=/208.67.222.222
NetGuard.Service: dns=/telcoDNS2
NetGuard.Service: dns=/telcoDNS1
Vpn     : Established by eu.faircode.netguard on tun0
NetGuard.Service: Start native log=false filter=true
NetGuard.Service: Hosts file use=true exists=false
NetGuard.Service: Forward protocol=17 port 53 to 127.0.0.1/5353 uid 9999
NetGuard.JNI: Starting tun 226 fwd53 1 level 5 thread 0
NetGuard.JNI: Started thread 9ee55000
NetGuard.JNI: Start events tun=226 thread 9ee55000
```

What does `clear` mean in `Stop tun`? Kill all connections I guess?

**Comment by M66B** [comment]
Idle means that the connection was idle to long (which is short, 15 seconds, for DNS to not run into too many files open).

Native restart is not a full restart, look for VPN restart.
I have tested this by switching from Wi-Fi to mobile and back and confirmed it works.
There might be an addional native restart as well.

Clear means terminate connections, which will never be done when reloading.

**Comment by licaon-kter** [comment]
> Native restart is not a full restart, look for VPN restart.

Yeah missed it, it was just above native.

So apparently these differences matter here.

Another thought, maybe it hits [this Samsung bug](https://github.com/schwabe/ics-openvpn/blob/3ebdb6482f3bab9423436e6b208963d5d7c07656/main/src/main/java/de/blinkt/openvpn/core/OpenVPNService.java#L590), say tries first DNS (custom) and then backtrack to system, although this does not explain why forwarding is ignored though. :(

**Comment by licaon-kter** [comment]
Looking at the way Samsung 5.0.2 (NetGuard-v2.0-2016022901-all-release) does a connectivity change (but which correctly keeps my DNS and forwarding):
```
NetGuard.Service: Legacy restart
NetGuard.Service: Stop native clear=false
NetGuard.JNI: Stop tun 63 clear 0 thread b94efac0
NetGuard.JNI: Kill thread b94efac0
NetGuard.JNI: Join thread b94efac0
NetGuard.JNI: pselect signaled tun 63 thread b94efac0
NetGuard.Service: Native exit reason=null
NetGuard.JNI: Stopped events tun=63 thread b94efac0
NetGuard.JNI: Stopped thread b94efac0
NetGuard.Service: Stopping
NetGuard.Service: vpn4=10.1.10.1 vpn6=fd00:1:fd00:1:fd00:1:fd00:1
NetGuard.Service: DNS system=telcoDNS1,telcoDNS2 VPN=208.67.220.220
NetGuard.Service: dns=/208.67.220.220
NetGuard.Service: dns=/telcoDNS1
NetGuard.Service: dns=/telcoDNS2
NetGuard.Service: Start native log=true filter=true
NetGuard.Service: Hosts file unchanged
NetGuard.Service: Forward protocol=17 port 53 to 127.0.0.1/5353 uid 10252
NetGuard.JNI: Starting tun 66 fwd53 1 level 5 thread 0
NetGuard.JNI: Started thread b84b1698
NetGuard.JNI: Start events tun=66 thread b84b1698
```

Two things stand out, first `Legacy restart` which is [here](https://github.com/M66B/NetGuard/blob/5dc1e66fbff76d82c275a01a33fa04326be7a944/app/src/main/java/eu/faircode/netguard/SinkholeService.java#L412), it reads that you don't try `seamless handover` in SDK<22? Although looking at [establish()](https://developer.android.com/reference/android/net/VpnService.Builder.html#establish()) says it was added in SDK14. Are you doing this for a bug or?

And second, I guess because of the first above, this section:
```
NetGuard.Service: vpn4=10.1.10.1 vpn6=fd00:1:fd00:1:fd00:1:fd00:1
NetGuard.Service: DNS system=telcoDNS1,telcoDNS2 VPN=208.67.220.220
NetGuard.Service: dns=/208.67.220.220
NetGuard.Service: dns=/telcoDNS1
NetGuard.Service: dns=/telcoDNS2
```
is here although it's not in SDK22 where I'm experiencing this issue (with reload) but it's there when i stop/restart. 

Also I don't see `NetGuard.Service: Stopping` on every connection change (with reload), but I do see it on stop/restart, is this something important?

OH WAIT you where editing that https://github.com/M66B/NetGuard/commit/e8f0cbb0421760e5bdecdb029d7afbb06db5c1ec while I was writing this... ummm... not sure if this will be still valid on a new build... you might even make it not work on the Samsung 5.0.2.

**Comment by M66B** [comment]
Android 5.0 has proven to be unsuitable for a VPN handover, even though the establish documentation says it should be possible, so there will be a (legacy) VPN restart with a pause between the stop and start for Android version 5.0 or lower.

I am surprise to see a DNS name 'telcoDNS1', which might be the problem, because there is no DNS to resolve the name of a DNS server. Do you have name resolving without setting a DNS server?

The commit is merely refactoring and will not change things.

Can you please post the output of:

```
adb shell getprop | grep -i dns
```

**Comment by licaon-kter** [comment]
> I am surprise to see a DNS name 'telcoDNS1'

No worries, just replaced IP for privacy :), NetGuard gets the correct IPs.

**Comment by M66B** [comment]
This version will route DNS explicitly into the VPN:

http://www.netguard.me/NetGuard-v2.0-2016022901-all-release-route.apk

I am not sure if I will release this, since I am afraid it could lead to other trouble.

**Comment by licaon-kter** [comment]
No joy.

While I retest to grab a full logcat, what are these two from .JNI?
* `UDP remove blocked session uid 1000`
* `UDP sendto error 1: operation not permitted`

**Comment by M66B** [comment]
Only the send error is relevant.
What is the target address?

licaon-kter <notifications@github.com> schreef op 29 februari 2016 15:45:17 CET:

**Comment by licaon-kter** [comment]
That's the whole error.
Above that there is a `UDP4 redirect to 127.0.0.1/5353` (aka the forwarding) but there are plenty of these in the logcat (as expected) although the local DNS is not actually used but TelCo's DNSes are.

**Comment by M66B** [comment]
Apparently your ROM does not allow sending to local host at all, even if the port is not a privilege port. This is likely the cause of the problem and unfortunately this cannot be fixed without root.

licaon-kter <notifications@github.com> schreef op 29 februari 2016 15:57:55 CET:

**Comment by licaon-kter** [comment]
Yes, that might be part of the problem since I see these (although DNS forwarding working as expected)
```
localdnsserver: type=1400 audit(0.0:31606): avc: denied { getopt } for laddr=127.0.0.1 lport=5353 scontext=u:r:init:s0 tcontext=u:r:init:s0 tclass=udp_socket op_res=0 ppid=1 pcomm="init" tgid=1 tgcomm="init"
localdnsserver: type=1400 audit(0.0:31607): avc: denied { getattr } for laddr=127.0.0.1 lport=5353 scontext=u:r:init:s0 tcontext=u:r:init:s0 tclass=udp_socket op_res=0 ppid=1 pcomm="init" tgid=1 tgcomm="init"
localdnsserver: type=1400 audit(0.0:31608): avc: denied { read } for laddr=127.0.0.1 lport=5353 scontext=u:r:init:s0 tcontext=u:r:init:s0 tclass=udp_socket op_res=0 ppid=1 pcomm="init" tgid=1 tgcomm="init"
localdnsserver: type=1400 audit(0.0:31609): avc: denied { write } for lport=37966 scontext=u:r:init:s0 tcontext=u:r:init:s0 tclass=udp_socket op_res=0 ppid=1 pcomm="init" tgid=1 tgcomm="init"
```
But since custom VPN DNS is not respected either (without a full VPN stop/restart first), there might be something else.

Other idea, TelCo gives my connection an IP of 10.xxx.xxx.xxx (from the logcat) although external is proper 211.xxx.xxx.xxx, the one thing involving that [is tethering](https://github.com/M66B/NetGuard/blob/e006d18feb45207a93f5f8c95ca20d6627a61429/app/src/main/java/eu/faircode/netguard/SinkholeService.java#L1010) but tried with it off too, no joy either.

For now, my best take in mind might be to force [this type of reload](https://github.com/M66B/NetGuard/blob/e006d18feb45207a93f5f8c95ca20d6627a61429/app/src/main/java/eu/faircode/netguard/SinkholeService.java#L413) for SDK22 too.

**Comment by M66B** [comment]
The problem with 'that' type of reload is that it is expensive (battery usage).

**Comment by licaon-kter** [comment]
Tethering support off, logcat as root, grep on `vpn` and `dns`: sequence [Wi-Fi->mobile->Wi-Fi](https://gist.github.com/licaon-kter/3ca3d44421bd1a5e5093)

`Attempting to register duplicate agent for type 17` looks odd.

**Comment by M66B** [comment]
This may be caused by the seamless VPN handover and may indicate a bug in the connectivity manager.

**Comment by licaon-kter** [comment]
Ok, so let's find the cause then, `forcestart` makes the VPN start/destroy itself at least 2 or 3 times on each connectivity switch which is rather unnerving anyway.
_(BTW, `forcerestart` was not needed on mobile->Wi-Fi switch though, just de other way around)_

More ideas?

**Comment by M66B** [comment]
The VPN should be fully restarted (with a seamless handover though) when the connectivity changes with the latest commits, so start by checking if this happens ("VPN restart").

**Comment by licaon-kter** [comment]
What was that about `VPN restart` exactly?

**Comment by licaon-kter** [comment]
https://github.com/M66B/NetGuard/commit/2c0d0bc7600ad45c924e72da3bd24eff992fa534 in not the cure. :(

**Comment by licaon-kter** [comment]
Although it was in front of my eyes all along, I'm still not train to think `it must be an Android bug` first but rather `I'm doing something wrong`.

Anyway, this https://code.google.com/p/android/issues/detail?id=35431 bares resemblance to what I see (although I don't understand the second post, it's pretty old anyway, API wise) .
Basically my TelCo has a proxy server/port set up in the APN and when the VPN is doing handover all requests are then sent to it instead, but if the VPN is destroyed&recreated the thing gets fixed.

OpenVPN for Android has just [a warning text](https://github.com/schwabe/ics-openvpn/blob/master/main/src/main/res/values/strings.xml#L192). 

Also of note https://code.google.com/p/android/issues/detail?id=33935 is not fixed, so I need to test if it will actually bypass the VPN or even *iptables*.
